[X/case] This refines guidance on usage of X and case statements

msfschaffner opened the pull request:
Based on our internal discussion this week, this PR refines the guidance on the usage of `unique case` statements and `X` literals in RTL. 

In particular, the guidance on `X` has been completely revised in order to favor a style that eliminates `X` literals in RTL as much as possible, and encourages the extensive use of SVAs to check invalid behavior in RTL instead. 

Note that this guidance shall not affect DV code that is not meant to be synthesized, and `X` can still be used if it is safe to do so.

Signed-off-by: Michael Schaffner <msf@google.com>

msfschaffner commented:
Note that this is a WIP PR. Once the initial draft has been stabilized, I will invite more reviewers to comment on the changes.

msfschaffner commented:
Should we mention somewhere in the style guide that `X` are still ok in DV related code that is not meant to be synthesized? I believe this could also solve some of @GregAC's concerns regarding `X` assignments used in bus models.

sjgitty commented:
> Should we mention somewhere in the style guide that `X` are still ok in DV related code that is not meant to be synthesized? I believe this could also solve some of @GregAC's concerns regarding `X` assignments used in bus models.

@sriyerg is considering (probably has already started) a style guide for verification code. We could put a brief mention here and save the rest for that section / document.

GregAC commented on VerilogCodingStyle.md:
I presume ASSERT_KNOWN is an assertion that a signal isn't X? Have we actually got this assertion defined anywhere? Even if we do we should define what it or point towards a definition in the style guide.

msfschaffner replied:
Yes, `ASSERT_KNOWN` is essentially what you sometimes refer to as `NEVER_X`. Good catch, I think we only have implicit references to these macros in the style-guide. I.e., if you search for `ASSERT`, you'll find another `ASSERT_INIT` and `ASSERT` reference, but these macros are never explained explicitly. I'll amend this and add a short section that explains these.

GregAC commented on VerilogCodingStyle.md:
Need to define what ASSERT_I is

msfschaffner replied:
Yep, see previous comment.

GregAC commented on VerilogCodingStyle.md:
```diff
@@ -1928,26 +2007,63 @@ always_comb begin
+  - if no default assignments are given before the case statement as shown in
+  the example above, any variables assigned in one case item must be assigned in
+  all case items, including the `default:`. Failing to do this can lead to a
+  simulation-synthesis mismatch as described in [Don Mills' paper][yalagp].
```
Link missing?

msfschaffner replied:
Oh yeah that got lost...

msfschaffner replied:
actually this is a cross reference within the document, so that link is correct.

msfschaffner commented on VerilogCodingStyle.md:
```diff
@@ -1928,26 +2007,63 @@ always_comb begin
+      state_d = Idle;
+      outc = in2;
+    end
+    default: ; // always assign a default, even if empty!
```
good point, I'll add.

msfschaffner commented on VerilogCodingStyle.md:
yes, will fix.

msfschaffner commented on VerilogCodingStyle.md:
```diff
@@ -1797,92 +1797,171 @@ increment.
+don't care conditions can lead to simulation/synthesis mismatches.
 
-Example:
+Instead of assigning and propagating `X` in order to flag and detect invalid
```
Indeed, I'll add that.

msfschaffner commented on VerilogCodingStyle.md:
fixed

msfschaffner commented on VerilogCodingStyle.md:
added

msfschaffner commented on VerilogCodingStyle.md:
thanks @sjgitty and @cdgori, I added this point to the list.

msfschaffner commented on VerilogCodingStyle.md:
```diff
@@ -1797,92 +1797,171 @@ increment.
-Example:
+Instead of assigning and propagating `X` in order to flag and detect invalid
+conditions, it is encouraged to make **extensive use of SVAs**. The added
+benefits of this design practice is that
```
thanks, fixed

msfschaffner commented on VerilogCodingStyle.md:
fixed

msfschaffner commented on VerilogCodingStyle.md:
```diff
@@ -1928,26 +2007,63 @@ always_comb begin
 The `unique` prefix is recommended before all case statements, as it creates
 simulation assertions that can catch certain mistakes. In some cases, `priority`
-may be used instead of `unique`, though `if-else-if` structures are a more
+may be used instead of `unique`, though in such cases, `if-else-if` structures
```
I replaced this with cascaded ternary structures

msfschaffner commented on VerilogCodingStyle.md:
Yeah I completely agree with your point, although I think that should probably go into our additional RTL coding guide for security critical applications.

msfschaffner replied:
WDYT, @sjgitty ?

GregAC replied:
If by large LUTs you mean large roms encoded with case statements or similar I don't think it's unreasonable to say you must use 0 or 1 for the bits you don't care about (designers choice, e.g. if you want to output 8'b1xxxxxxx then 1 may be the sensible choice as it matches the other bit or if there's lot of inputs that are total don't cares on output then try to match the most used bit say you output either one or two 1s with 0s and otherwise Xs the don't care cases would become zeros).

You could also say that if you end up with a 'sparse LUT' (i.e. one where we don't care about much of the inputs and/or outputs) then you may be doing it wrong and should rethink.

As a final fallback recommend you push the LUT with Xs through some logic minimization tool to get the same LUT back but with choices for the Xs that work well (though I'm not sure such a tool exists you'll just get a bunch of gates and we may get back to the same information leakage issue).

msfschaffner replied:
Indeed, I meant variations of large case statements or arrays.

In the past, people have been using tools like [Espresso](https://en.wikipedia.org/wiki/Espresso_heuristic_logic_minimizer) to minimize large truth tables - although in my own experiments I got the best results by specifying large LUTs with arrays in SystemVerilog and running them straight through DC. 

I never really specified sparse LUTs in this manner, but I did use truncated arrays for unaligned LUTs. I.e., say you have an LUT with 567 words, I would use an array `[0:566]`, which implicitly means that `[567:1023]` are `X` when indexed with higher indexes. But then again padding with zero up to the next pow2 depth, or using a guarding if statement around the indexing operation would likely not add much more logic and would make this predictable and safe.

So I would say, let's make this strict and just remove that statement on LUTs.

msfschaffner commented on VerilogCodingStyle.md:
```diff
@@ -1928,26 +2003,65 @@ always_comb begin
+  all case items, including the `default:`. Failing to do this can lead to a
+  simulation-synthesis mismatch as described in [Don Mills' paper][yalagp].
+
+The following is a different example showing a style-compliant case statement
```
@sjgitty shall I omit this second case statement example? looks like this is not strictly needed after stating the points above...

sjgitty replied:
I think it is fine as written now, thanks.

msfschaffner commented on VerilogCodingStyle.md:
do you think we should remove this note concerning the assertion macros?

msfschaffner replied:
ah and II have added the note you and @cdgori suggested in the section on X-avoidance.

msfschaffner commented:
Added a revised version of the style guide. 

@mwbranstad, I also added an example that shows which operator to use within the context of flip-flop resets `if (!rst_ni)` as opposed to `if (~rst_ni)`.

imphil commented:
Can you do the second commit in a separate PR? That one should be easy to agree on and get merged, and will keep the discussion here focused.

GregAC commented on VerilogCodingStyle.md:
```diff
@@ -1797,92 +1797,167 @@ increment.
+
+To comply with this style, RTL must place `` `ASSERT_KNOWN`` assertions on all
+module outputs, with the exception of signals that may implicitly be `X` at the
+beginning of the simulation, such as FIFO, SRAM or register file outputs.
```
In the separate security guidelines we way want to say all such structures should receive an explicit reset to 0 via FSM during an init phase?

msfschaffner replied:
That is another question for @cdgori :)...

msfschaffner commented:
Removed the sentence stating that large LUTs may be an exception to the new guidance on `X`.

msfschaffner commented:
Removed the following recommended example, as it uses an unclocked assertion which may be susceptible to glitches in simulation:
```systemverilog
// alternative using inline assertion
always_comb begin
  out0 = '0;
  out1 = '0;
  unique case (sel)
    mode1: out0 = foo;
    mode2: out1 = bar;
    default: `ASSERT_I(Invalid_A, 0);
  endcase
end
```
See also https://github.com/lowRISC/ibex/pull/528 for context.

sjgitty commented on VerilogCodingStyle.md:
```diff
@@ -1928,26 +1989,65 @@ always_comb begin
+  outb = 1'b0;
+  outc = 1'b0;
+
+  unique case  (state_q)
```
nit: double-space here

msfschaffner replied:
thanks, fixed.

msfschaffner commented:
I am putting this up for review after this initial round of comments. PTAL.

We are likely going to do a general editing pass in the future to make the style guide more compact - but this will be a separate PR.

msfschaffner commented:
Squashed.

msfschaffner commented:
Are there any other comments that should be addressed before moving this one in?

@GregAC can you give this another look and unblock the PR if you are happy with the current state?

sjgitty commented on VerilogCodingStyle.md:
```diff
@@ -1928,26 +1989,65 @@ always_comb begin
+
+The following is a different example showing a style-compliant case statement
+variant that is frequently used for describing the next-state logic of a finite
+sate machine. What is different from the previous example is that the default
```
typo: ^sate^state^

msfschaffner replied:
fixed, thanks

sjgitty commented on VerilogCodingStyle.md:
Personally I think we can skip this section until we have full guidance here. But let's go ahead and leave it here, but please add something like "More security assertion and coding style guidance will be coming" else this seems like an odd incomplete story on the topic.

msfschaffner replied:
Ok I added the following sentence at the end of this subsection:
```
More security assertion and coding style guidance will be given in a separate
document, soon.
```

msfschaffner commented:
> A few comments, but otherwise this is looking good to me. If there isn't an issue already filed, please file one to create security-related coding style and assertion guidance, CCing the usual suspects.

Thanks for giving it another look, I've filed an issue.

sriyerg commented:
> > Should we mention somewhere in the style guide that `X` are still ok in DV related code that is not meant to be synthesized? I believe this could also solve some of @GregAC's concerns regarding `X` assignments used in bus models.
> 
> @sriyerg is considering (probably has already started) a style guide for verification code. We could put a brief mention here and save the rest for that section / document.

@udinator, let's ensure that we address this in the DV style guide.

msfschaffner commented:
Ok it looks like we converged on this one - merging.
